### ISSUE TITLE
support GraphQL enum types

### ISSUE THREAD

**Issue by maxpain** [hasura/graphql-engine#982]
Why Hasura is not utilizing Postgres ENUM types and not translate it to GraphQL Enums?

**Comment by shahidhk** [comment]
We do not support Postgres ENUM types as of now. Will mark this as a feature request, give a :+1: if you'd like the feature added. :smile: 

Till then, you can use [single-column tables](https://docs.hasura.io/1.0/graphql/manual/schema/enums.html)

**Comment by 0x777** [comment]
We had a discussion about this couple of days. We can go ahead with the implementation with some feedback from the community. 

We understand that it is incredibly useful to propagate enums in your database to the GraphQL layer. There are couple of ways in which you typically model enums in the database.
1. Using [Postgres enums](https://www.postgresql.org/docs/current/datatype-enum.html): There are certain limitations with Postgres's enums, the most important one is that you can't delete a value from an enum.
2. The other option is to to use single-column tables for enums as documented [here](https://docs.hasura.io/1.0/graphql/manual/schema/enums.html). This [some](https://www.slideshare.net/billkarwin/sql-antipatterns-strike-back/82-ENUM_Antipattern_Antipattern_use_ENUM) consider is a better practice over using Postgres enum types. 

We would like to know which pattern is commonly used so that we can do the appropriate propagation to the graphql layer.

**Comment by chunsli** [comment]
Is there a way to implement a text array, where each value is one of the enumerations. is there a way of doing this without a normalized table?

**Comment by hmexx** [comment]
+1 for the second pattern @0x777

**Comment by lordofthelake** [comment]
I think it depends on the use case. For values that are likely to change over time, a separate table is better. 

But for values that are "set in stone" (the examples that come to mind are static data like country codes, currencies, language codes, measurement units and the likes) a native Postgres enum would be _very_ valuable (and I would argue, also a better practice – more efficient, less error prone).

**Comment by frankdugan3** [comment]
In my opinion, there is a lot of value in equating Postgres enums with GraphQL enums. Neither should change often, and the benefit is great documentation for the API. Using the single-column table pattern in this case would move the error for selecting an invalid value from something that could be checked statically at compile time to something only detected at runtime.

If the value is more flexible or even user defined, I think the single-column table pattern makes sense.

**Comment by drewwestphal** [comment]
Another issue to consider in supporting this feature is the migration side. As you can see from the below, adding values to an enum type is not possible within "transactional migrations" (or transactions generally)

https://stackoverflow.com/questions/1771543/adding-a-new-value-to-an-existing-enum-type/10404041#10404041

**Comment by respectTheCode** [comment]
Looks like this might be a blocker for us. We use ENUMs in our existing database and I am getting an error stating that my schema is invalid from the GraphQL CLI.

**Comment by respectTheCode** [comment]
false alarm. my problem was related to having a row permission with no columns selected. once I fixed that the schema worked.

**Comment by Krever** [comment]
Hey, this is also a problem with codegen tools, that relies on schema to generate proper enums. Right now possible enum values cannot be detected by such tools as postgres enums are translated to grpahql scalar types.

**Comment by mnlbox** [comment]
@Krever why you use third party schema generator when we have console app for Hasura :thinking: 
Any use-case?

**Comment by Krever** [comment]
@mnlbox I may not phrased myself clearly enough. Im using codegen tool to generate scala code which will be used as a hasura client. The codegen relies on schema returned by hasura via introspection.

**Comment by Fxlr8** [comment]
any updates on this one?

**Comment by respectTheCode** [comment]
any updates?

**Comment by dsandip** [comment]
@respectTheCode @Fxlr8 Work on this issue has begun. Please expect it to land in one of the next two releases.

**Comment by mnlbox** [comment]
Now I have many redundant tables and relations because of the lack of this feature. @dsandip It's an awesome news that Hasura can handle this for me and I can remove many tables and have a clean data diagram.
Thanks guys :wink:

**Comment by dsandip** [comment]
@mnlbox That's great to hear! 

Also, clarifying that the plan is to implement enums using single-column tables as outlined by Vamshi [here](https://github.com/hasura/graphql-engine/issues/982#issuecomment-457100594).

**Comment by Krever** [comment]
@dsandip what about postgres enums that we already have? For cases we want to have control over database and only serve it with hasura.

**Comment by dsandip** [comment]
@Krever You can continue using Postgres Enums the way you are using them now. The single-column table approach will just create a new table for every enum and associate it with your tables that need this enum's values. You can modify these tables yourself too i.e. you'll have control over them. Does this meet your requirements? If not, could you please elaborate on your use-case.

**Comment by Krever** [comment]
Hmm, so it means that if I have a defined enum type and a table using it it will just work? 
Example:
```
CREATE TYPE mood AS ENUM ('sad', 'ok', 'happy');
CREATE TABLE person (
    name text,
    current_mood mood
);
```
And I will be able to work with such table via graphql?  Then its alright! Although I dont get the purpose of this new single-column table.

**Comment by dsandip** [comment]
@Krever 

>  if I have a defined enum type and a table using it it will just work? And I will be able to work with such table via graphql?

Yes, you can access/modify this table via graphql even now. Here's a sample of an insert mutation:

```graphql
mutation {
  insert_person(objects: {name: "sandip", current_mood: "ok"}) {
    affected_rows
    returning {
      id
      current_mood
      name
    }
  }
}
```
Response:
```JSON
{
  "data": {
    "insert_person": {
      "affected_rows": 1,
      "returning": [
        {
          "id": 1,
          "current_mood": "ok",
          "name": "sandip"
        }
      ]
    }
  }
}
```

This single-column table will be used to support GraphQL enums (*instead of using Postgres enums for this*).

**Comment by Krever** [comment]
Yes, this I'm aware of. I was hoping that my postgres enums will be expressed as graphql enums in schema.

**Comment by mnlbox** [comment]
@dsandip some tools like react-admin have some limitation and convention for this single-column table. Also I think maybe it's better to prevent this type of table by default in console app. For example we user can show/hide this type of table with an environment variable. Because I have many enum tables and want to make my console clean. :wink:

**Comment by respectTheCode** [comment]
Does that PR add support for Postgres ENUMs? The description seems to describe a different feature.

**Comment by Kjens93** [comment]
Note: It looks like Postgres 12 will include support for adding enum values within a transaction. https://www.postgresql.org/docs/12/sql-altertype.html

**Comment by Kjens93** [comment]
@respectTheCode:
> Does that PR add support for Postgres ENUMs? The description seems to describe a different feature.
From the looks of it, it appears that #2672 does **not** add support for exposing Postgres enums as Graphql enums. It appears that this issue should be reopened since the supposed "fix" does not actually accomplish what was requested.

**Comment by Kjens93** [comment]
Note: It looks like Postgres 12 will include support for adding enum values within a transaction. https://www.postgresql.org/docs/12/sql-altertype.html

**Comment by NSilv** [comment]
This is a dealbreaker for us as well. We have a lot of enum values sitting around that we really don't want to change into tables for a lot of reasons. Lack of support basically means that we have to instead make endpoints (or alternatively actions) for each field that we want to update (we can read the enum values just fine, but we can't filter by them in where clauses, nor can we run mutations on them). Is there any chance this is getting implemented at some point?

**Comment by respectTheCode** [comment]
We moved on and built our graphql backend with Apollo Federation. It turned out to be a lot less work than we had expected and will be a better long term decision too.

**Comment by mnlbox** [comment]
Happy to see this is reopened again :wink: 
Really waiting for this feature.

**Comment by kasugaicrow** [comment]
> @respectTheCode @Fxlr8 Work on this issue has begun. Please expect it to land in one of the next two releases.

not to be rude but do we have any news today? I just cant add new tables for each enums in my schema.

**Comment by kasugaicrow** [comment]
nvm

**Comment by factoidforrest** [comment]
I'm just so confused why Hasura went to the trouble of implementing their "referenced table enums" before they tried to make native enums work. Graphql has enums, postgres has enums...as far as I can tell, they're almost a 1:1.  

Yes, Enums should not be used for dynamic data.  That doesn't mean you should NEVER use them and not support them.  

Maybe this is an aside, but it feels similar to the lack of support for arrays, right down to having to insert arrays as psql formatted strings. Yes, arrays aren't always the right abstraction, but does that mean they should be totally unsupported and broken? Just because you've seen something abused doesn't mean it shouldn't exist.

**Comment by JuSacco** [comment]
Please consider adding this feature as at least as experimental.

**Comment by maxpain** [comment]
Any updates?

**Comment by ciphrd** [comment]
I find it concerning that after 5 years this hasn't been solved properly. As described by others, there are many good reasons to use native enums as opposed to single column tables, speed and storage optimisations being the main ones (there is a significant impact on tables with 1M+ rows, the different between having enums and string can turn into 100's of MBs of savings).

IMO an aggregator such as Hasura shouldn't take such opinionated design decisions on what are and what aren't good DB practices, especially when in this case all approaches have benefits.

The biggest impact of this decision for us is that our codegen tools don't type many of our enums at all, which impacts DX and code safety by a great margin (also prevents various type-safe design patterns of being implemented).

I think it would be great to add support for both cases.
